Add RangeFieldMapper for numeric and date range types

nknize opened the pull request:
Lucene 6.2 added support to index and query numeric ranges. This PR adds a new `RangeFieldMapper` for indexing numeric (int, long, float, double) and date ranges and creating appropriate range and term queries. The design is similar to `NumericFieldMapper` in that it uses a `RangeType` enumerator for implementing logic specific to each numeric type. The following new field types are supported by this mapper: `int_range`, `float_range`, `long_range`, `double_range`, `date_range`.

Lucene does not provide a DocValue field specific to RangeField types so the `RangeFieldMapper` implements a `CustomRangeDocValuesField` for handling doc value support.

When executing a Range query over a Range field, the `RangeQueryBuilder` has been enhanced to accept a new `relation` parameter for defining the type of query as one of: `WITHIN`, `CONTAINS`, `INTERSECTS`. This provides support for finding all ranges that are related to a specific range in a desired way. As with other spatial queries, `DISJOINT` can be achieved as a `MUST_NOT` of an `INTERSECTS` query.

Unit and Integration tests are provided.

TODO:
- add documentation for new field type 

closes #20999

nknize commented:
Thanks for the great feedback @jpountz. I'l fix up the formatting changes. Looks like IntelliJ tried to be to "intelligent" for me. :)

+1 on the other suggestions as well. DocValues can wait until they're needed by aggregations, and `ignore_malformed` and default null_values were more of a pain. I was never really happy with the naming `from_null_value` `to_null_value` anyway. I'll make the changes and finish adding documentation.

jpountz commented on docs/reference/mapping/types/range.asciidoc:
```diff
@@ -0,0 +1,104 @@
+  }
+}
+--------------------------------------------------
+// CONSOLE
```
can you add a TESTRESPONSE block?

jpountz commented on docs/reference/mapping/types/range.asciidoc:
```diff
@@ -0,0 +1,104 @@
+<1> `date_range` types accept the same field parameters defined by the <<date, `date`>> type.
+<2> Example indexing a meeting with 10 to 20 attendees.
+<3> Date ranges accept the same format as described in <<query-dsl-range-query, date range queries>>
+<4> Example date range using date time stamp. This also accepts <<date-math, date math>> formatting.
```
what happens if `now` is used?

jpountz commented on docs/reference/mapping/types/range.asciidoc:
```diff
@@ -67,6 +68,8 @@ GET my_index/_search
 <4> Example date range using date time stamp. This also accepts <<date-math, date math>> formatting.
 <5> Cost of the meeting ranges from 60 to 99.99 currency.
 <6> Range queries work the same as described in <<query-dsl-range-query, range query>>
+<7> Range queries over range fields support a `relation` parameter which can be one of `WITHIN`, `CONTAINS`,
```
maybe add a link to the mappings docs

clintongormley commented on docs/reference/mapping/types/range.asciidoc:
```diff
@@ -0,0 +1,107 @@
+  }
+}
+--------------------------------------------------
+// TESTRESPONSE
```
You should have a `// CONSOLE` here, and the `// TESTRESPONSE` block should go under another source code block which contains the results of the query.  This block can be inside a comment block as below so that it doesn't render in the docs, eg:

    //////////

    [source,js]
    ----------
    { 
      "took": 25,
      "hits": ...
    }
    ----------
    // TESTRESPONSE[s/"took": 25/"took": $body.took/]

    //////////

clintongormley commented:
I love how you've integrated this feature - awesome!

I presume you can specify multiple ranges in a single field?  Could you add an example to the docs which shows how to do this?

nknize commented:
OK, I've addresses all change requests:

* removed `ignore_malformed`
* removed `doc_value` support (for now)
* removed `null_value` support
* added support for `now`
* Fixed up docs
* added `TESTRESPONSE` to doc testing

jpountz commented on core/src/main/java/org/elasticsearch/index/mapper/RangeFieldMapper.java:
```diff
@@ -0,0 +1,735 @@
+import static org.elasticsearch.index.query.RangeQueryBuilder.GTE_FIELD;
+import static org.elasticsearch.index.query.RangeQueryBuilder.LT_FIELD;
+import static org.elasticsearch.index.query.RangeQueryBuilder.LTE_FIELD;
+//import static org.elasticsearch.index.query.RangeQueryBuilder.TIME_ZONE_FIELD;
```
Let's just remove those imports instead of commenting them out?

jpountz commented on core/src/main/java/org/elasticsearch/index/mapper/RangeFieldMapper.java:
```diff
@@ -0,0 +1,735 @@
+                Object propNode = entry.getValue();
+                if (propName.equals("null_value")) {
+                    throw new MapperParsingException("Property [null_value] is not supported for [" + this.type.name
+                            + "] field types. Use [from_null_value] or [to_null_value] instead.");
```
This looks outdated?

jpountz commented on core/src/main/java/org/elasticsearch/index/mapper/RangeFieldMapper.java:
```diff
@@ -0,0 +1,735 @@
+            setupFieldType(context);
+            return new RangeFieldMapper(name, fieldType, defaultFieldType, coerce(context), includeInAll,
+                context.indexSettings(), multiFieldsBuilder.build(this, context), copyTo);
+        }
```
Since doc values are not supported yet, I think we'd need something that fails if a user tries to enable them?

nknize replied:
++ done

jpountz commented on core/src/test/java/org/elasticsearch/index/mapper/RangeFieldTypeTests.java:
```diff
@@ -0,0 +1,166 @@
+    protected RangeFieldMapper.RangeFieldType createDefaultFieldType() {
+        return new RangeFieldMapper.RangeFieldType(type);
+    }
+
```
you should override `setupProperties` to add teh custom properties of this field type such as the date formatter so that they are tested, you can have a look at DateFieldTypeTests for instance.

nknize replied:
++ done

jpountz commented on core/src/main/java/org/elasticsearch/index/mapper/RangeFieldMapper.java:
```diff
@@ -0,0 +1,735 @@
+            if (!locale.equals(dateTimeFormatter.locale())) {
+                fieldType().setDateTimeFormatter(new FormatDateTimeFormatter(dateTimeFormatter.format(),
+                    dateTimeFormatter.parser(), dateTimeFormatter.printer(), locale));
+            }
```
Can we fail if the range type is not a date but a date time formatter is set?

jpountz commented on core/src/main/java/org/elasticsearch/index/mapper/RangeFieldMapper.java:
```diff
@@ -0,0 +1,735 @@
+            if (!super.equals(o)) return false;
+            RangeFieldType that = (RangeFieldType) o;
+            return Objects.equals(rangeType, that.rangeType)
+                && Objects.equals(dateTimeFormatter, that.dateTimeFormatter);
```
is it on purpose that we do not take the timezone into account?

jpountz replied:
hmm maybe it is not even used?

jpountz commented on core/src/main/java/org/elasticsearch/index/mapper/RangeFieldMapper.java:
```diff
@@ -0,0 +1,735 @@
+                                name() + "], with unknown parameter [" + fieldName + "]");
+                        }
+                    }
+                } // while
```
remove comment?

jpountz commented on core/src/main/java/org/elasticsearch/index/mapper/RangeFieldMapper.java:
```diff
@@ -0,0 +1,735 @@
+        if (includeDefaults
+                || fieldType().dateTimeFormatter().locale() != Locale.ROOT) {
+            builder.field("locale", fieldType().dateTimeFormatter().locale());
+        }
```
If we plan on keeping the time zone, I think we should add it here.

jpountz commented on core/src/main/java/org/elasticsearch/index/mapper/RangeFieldMapper.java:
```diff
@@ -0,0 +1,735 @@
+            }
+            @Override
+            public Double maxValue() {
+                return Double.MAX_VALUE;
```
should these be NEGATIVE/POSITIVE_INFINITY instead?

jpountz replied:
(same question for FLOAT)

jpountz commented on core/src/test/java/org/elasticsearch/index/mapper/RangeFieldMapperTests.java:
```diff
@@ -0,0 +1,325 @@
+    @Override
+    protected void doTestIgnoreMalformed(String type) throws IOException {
+        // not supported for RangeFields
+    }
```
then could we only have this method in the test case for non-range numerics?

jpountz commented on core/src/test/java/org/elasticsearch/index/mapper/RangeFieldMapperTests.java:
```diff
@@ -0,0 +1,325 @@
+            .field(getFromField(), getFrom(type))
+            .nullField(getToField())
+            .endObject()
+            .endObject().bytes());
```
can we also test a `null` from?

nknize replied:
done. The first check tests null for both from and to.

jpountz commented on core/src/main/java/org/elasticsearch/index/mapper/RangeFieldMapper.java:
```diff
@@ -0,0 +1,735 @@
+                        }
+                    }
+                } // while
+                range = new Range(from, to, includeFrom, includeTo);
```
I think we should either raise an exception if `from` or `to` is null, meaning that one of the bounds was not specified, or make their values default to the min/max values of the range type. I think the latter is better as it is more consistent with the range query?

jpountz commented on core/src/test/java/org/elasticsearch/index/mapper/RangeFieldMapperTests.java:
```diff
@@ -0,0 +1,325 @@
+        assertTrue(storedField.fieldType().stored());
+        assertThat(storedField.stringValue(), containsString(type.equals("date_range") ? "1477872000000" : "5"));
+        assertThat(storedField.stringValue(), containsString(getMax(type) + ""));
+    }
```
I'd also like a test for the case that either a single bound or none of the bounds are specified (even if that means checking that an exception is thrown depending on the decision we make).

nknize commented:
Thanks for the review @jpountz I like the suggestions for better exception handling. I've addressed all changes and it should be ready for another round of review.

jpountz commented on core/src/main/java/org/elasticsearch/index/mapper/RangeFieldMapper.java:
```diff
@@ -0,0 +1,777 @@
+    public static final class RangeFieldType extends MappedFieldType {
+        protected RangeType rangeType;
+        protected FormatDateTimeFormatter dateTimeFormatter;
+        protected DateTimeZone timeZone;
```
Is this parameter actually used somewhere? If yes we should copy it in the copy constructor, otherwise let's remove it?

jpountz commented on core/src/main/java/org/elasticsearch/index/mapper/RangeFieldMapper.java:
```diff
@@ -0,0 +1,777 @@
+            }
+            return fields;
+        }
+        public Number parse(RangeFieldType fieldType, XContentParser parser, boolean coerce, LongSupplier nowInMillis) throws IOException {
```
this method seems to be only used at indexing time, so I don't think it should accept `nowInMillis` since index dates need to be concrete dates rather than math expressions

jpountz commented on core/src/main/java/org/elasticsearch/index/query/RangeQueryBuilder.java:
```diff
@@ -123,6 +132,11 @@ protected void doWriteTo(StreamOutput out) throws IOException {
+        if (this.relation != null) {
+            relationString = this.relation.getRelationName();
+        }
+        out.writeOptionalString(relationString);
```
you should protect it with a version check

jpountz commented on core/src/main/java/org/elasticsearch/index/mapper/RangeFieldMapper.java:
```diff
@@ -0,0 +1,777 @@
+        }
+        public Number parse(Object value, Number nullValue, FormatDateTimeFormatter dateTimeFormatter) throws IOException {
+            return value == null ? nullValue : numberType.parse(value);
+        }
```
can you add some minimal javadocs to these methods to know how they are used?

jpountz commented on core/src/main/java/org/elasticsearch/index/mapper/RangeFieldMapper.java:
```diff
@@ -0,0 +1,777 @@
+            @Override
+            public Number parse(RangeFieldType fieldType, XContentParser parser, boolean coerce, LongSupplier nowInMillis)
+                    throws IOException {
+                return fieldType.dateMathParser.parse(parser.text(), nowInMillis);
```
should we round the lower bound down and the upper bound up like DateFieldType.innerRangeQuery?
Also I think you could use a `nowInMillis` impl that just throws an exception since `now` is not supposed to be used at index time.

nknize commented:
Thanks @jpountz! I've made the next set of corrections and are ready for a (I think final?) review.

jpountz commented on core/src/main/java/org/elasticsearch/index/query/RangeQueryBuilder.java:
```diff
@@ -108,6 +114,12 @@ public RangeQueryBuilder(StreamInput in) throws IOException {
         if (formatString != null) {
             format = Joda.forPattern(formatString);
         }
+        if (in.getVersion().onOrAfter(Version.V_5_1_0_UNRELEASED)) {
```
Given the timing of this PR, this will rather be V_5_2_0

nknize replied:
++
